// ember-this-fallback/lib/helpers/deprecations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deprecationOptionsFor = void 0;
const assert_1 = __importDefault(require("../types/assert"));
const CURRENT_DEPRECATIONS = [
  [
    'this-property-fallback',
    {
        id: 'ember-this-fallback.this-property-fallback', // Updating deprecation id so we can distinguish it from Ember's
        until: 'n/a',
        for: 'ember-this-fallback',
        url: 'https://deprecations.emberjs.com/v3.x#toc_this-property-fallback',
        since: {
            available: '0.2.0',
        },
    },
  ]
];
const DEPRECATION_OPTIONS_MAP = new Map(CURRENT_DEPRECATIONS);
function deprecationOptionsFor(id) {
    const options = DEPRECATION_OPTIONS_MAP.get(id);
    (0, assert_1.default)(`expected to find DeprecationOptions for id=${id}`, options);
    return options;
}
exports.deprecationOptionsFor = deprecationOptionsFor;
